File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.py
k, n = map(int, input().split())
cm_lst = []

for _ in range(k):
   lans = int(input())
   cm_lst.append(lans)

left, right = 1, max(cm_lst)
   
while left <= right:
   mid = (left + right) // 2
   cnt = 0
   
   for i in cm_lst:
      cnt += i // mid
      #print(i, cnt, mid, left, right)
   
   if cnt >= n:
      left = mid + 1
   else:
      right = mid - 1
      
print(right)

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.io.*;
import java.util.*;

public class Main {

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        StringTokenizer st = new StringTokenizer(br.readLine());
        int K = Integer.parseInt(st.nextToken());
        int N = Integer.parseInt(st.nextToken());

        int[] lan = new int[K];
        int maxLan = 0;
        for (int i = 0; i < K; i++) {
            lan[i] = Integer.parseInt(br.readLine());
            if (lan[i] > maxLan) maxLan = lan[i];
        }

        long left = 1;
        long right = maxLan;
        long result = 0;

        while (left <= right) {
            long mid = (left + right) / 2;

            long count = 0;
            for (int l : lan) {
                count += (l / mid);
            }

            if (count >= N) {
                // 더 많이 쪼갰을 경우 lan 길이를 늘려야함, N개 이상 만들 수 있기 때문에 더 만들 수 있는지 탐색
                result = mid; // 정답 동기화
                left = mid + 1;
            } else {
               // 적을 경우 lan을 더 잘게 쪼개야함
               right = mid - 1;
            }
        }

        bw.write(result + "\n");
        bw.flush();
    }
}   
